### ISSUE TITLE
Cargo should skip crates.io check with --locked and all dependencies already resolved

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/cargo#10984]

**Comment by ehuss** [comment]
Transferred to the cargo repo.

> (Please let me know if I'm missing something that cargo needs internet access for if all dependencies are present and account for.)

It is checking if any dependencies have been yanked.

I'm not sure what might be a better way to handle that. At least it could provide a better error message, to perhaps suggest using `--offline`.

Copying the current error message:

```
    Updating crates.io index
warning: spurious network error (2 tries remaining): failed to connect to github.com: Connection refused; class=Os (2)
warning: spurious network error (1 tries remaining): failed to connect to github.com: Connection refused; class=Os (2)
error: failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to connect to github.com: Connection refused; class=Os (2)
```

**Comment by mqudsi** [comment]
* Can (should?) `--locked` be interpreted as allowing yanked dependency checks to be skipped? (I'm not sure if `cargo` with `--locked` specified currently warns about a yanked dependency or if it's a hard error, although I thought it was only the former.)
* Depending on the reason for checking crates.io, there may be no reason to error out at all (e.g. checking for yanked dependencies isn't _required_ to build a crate, whereas connecting to crates.io to download a missing dependency obviously would be).
* Can the timeout for connecting to `crates.io` be reduced in all such non-essential cases? That would help in general, as the OS default can be rather high (Windows default is 72s, iirc). (When testing/mocking, make sure not to just refuse/reject the outgoing connection but rather silently drop the outgoing packets as the behavior differs.)
* In general, a sane recency check for `cargo install` (even without `--locked`) prior to updating the crates.io index might not be a bad idea. Essential crates.io usages will continue to establish a connection to crates.io in order to download missing dependencies, but the worst case scenario here is that a just-updated crate will use a slightly older version (5 minutes stale? an hour?).

**Comment by epage** [comment]
> Can (should?) --locked be interpreted as allowing yanked dependency checks to be skipped? (I'm not sure if cargo with --locked specified currently warns about a yanked dependency or if it's a hard error, although I thought it was only the former.)

I believe its there to report warnings.

However, re-evaluating how high of fidelity the warnings have to be is reasonable.

Howver, at least in my testing, the scope of performance improvement, when the network is working, is small.  I did some quick tests on my machine with `--locked --offline` for `check` and `install`.  Maybe due to the synthetic nature of the test, I missed some cases though.

**Comment by mqudsi** [comment]
Yeah, in the happy case it's not too bad of a cost at all. The problem is the variability of the highly volatile nature of establishing network connections to third party servers as compared to all the other things that `cargo` does locally - there are just so many factors that could turn that 250-500ms overhead into a 2-5s overhead or a 20-50s overhead depending on your local environment (connection latency, packet loss), highly variable dns resolution timings (exacerbated by the variances resulting from the choice of dns server), and the load at the destination. GitHub has great uptime, but nothing's ever perfect. 

(I'm not clear on how the internals of the update check are performed, can a git gc run ever delay this as well?)

**Comment by epage** [comment]
I feel I should add that, in my evaluation, I have a high quality fiber connection and so the results could be dramatically worse for regions that will have significantly worse internet connections.

**Comment by epage** [comment]
Note: we have an unstable no-index-update feature (#7404)

**Comment by mqudsi** [comment]
Thanks, Ed!

**Comment by RalfJung** [comment]
> Howver, at least in my testing, the scope of performance improvement, when the network is working, is small.

That's a big "when". When you're developing on a train with spotty internet, latency can vary between 100ms now and 10s a bit later. ;)
